File: src/request_variables.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const types = @import("types.zig");
const RequestVariable = types.RequestVariable;
const RequestContext = types.RequestContext;
const HttpResult = types.HttpResult;

pub fn parseRequestVariable(allocator: Allocator, reference: []const u8) !RequestVariable {
    // Parse syntax: {{<request_name>.(request|response).(body|headers).(*|JSONPath|XPath|<header_name>)}}

    // Remove surrounding {{ and }}
    var cleaned = reference;
    if (std.mem.startsWith(u8, reference, "{{") and std.mem.endsWith(u8, reference, "}}")) {
        cleaned = reference[2 .. reference.len - 2];
    }

    var parts = std.mem.splitSequence(u8, cleaned, ".");

    const request_name = parts.next() orelse return error.InvalidRequestVariable;
    const source_str = parts.next() orelse return error.InvalidRequestVariable;
    const target_str = parts.next() orelse return error.InvalidRequestVariable;
    const path = parts.rest();

    const source = if (std.mem.eql(u8, source_str, "request"))
        RequestVariable.RequestVariableSource.request
    else if (std.mem.eql(u8, source_str, "response"))
        RequestVariable.RequestVariableSource.response
    else
        return error.InvalidRequestVariable;

    const target = if (std.mem.eql(u8, target_str, "body"))
        RequestVariable.RequestVariableTarget.body
    else if (std.mem.eql(u8, target_str, "headers"))
        RequestVariable.RequestVariableTarget.headers
    else
        return error.InvalidRequestVariable;

    return RequestVariable{
        .reference = try allocator.dupe(u8, reference),
        .request_name = try allocator.dupe(u8, request_name),
        .source = source,
        .target = target,
        .path = try allocator.dupe(u8, path),
    };
}

pub fn extractRequestVariableValue(allocator: Allocator, request_var: RequestVariable, context: []const RequestContext) !?[]const u8 {
    // Find the request context by name
    var target_context: ?*const RequestContext = null;
    for (context) |*ctx| {
        if (std.mem.eql(u8, ctx.name, request_var.request_name)) {
            target_context = ctx;
            break;
        }
    }

    if (target_context == null) {
        return null; // Request not found or not executed yet
    }

    const ctx = target_context.?;

    switch (request_var.source) {
        .request => return extractFromRequest(allocator, request_var, &ctx.request),
        .response => {
            if (ctx.result) |*result| {
                return extractFromResponse(allocator, request_var, result);
            } else {
                return null; // Request not executed yet
            }
        },
    }
}

fn extractFromRequest(allocator: Allocator, request_var: RequestVariable, request: *const types.HttpRequest) !?[]const u8 {
    switch (request_var.target) {
        .body => {
            if (request.body) |body| {
                if (std.mem.eql(u8, request_var.path, "*")) {
                    return try allocator.dupe(u8, body);
                }
                // For request body, we don't support JSONPath/XPath parsing yet
                // Return the whole body for any path (including non-wildcard)
                return try allocator.dupe(u8, body);
            }
            return null;
        },
        .headers => {
            for (request.headers.items) |header| {
                if (std.ascii.eqlIgnoreCase(header.name, request_var.path)) {
                    return try allocator.dupe(u8, header.value);
                }
            }
            return null;
        },
    }
}

fn extractFromResponse(allocator: Allocator, request_var: RequestVariable, result: *const HttpResult) !?[]const u8 {
    switch (request_var.target) {
        .body => {
            if (result.response_body) |body| {
                if (std.mem.eql(u8, request_var.path, "*")) {
                    return try allocator.dupe(u8, body);
                }

                // Basic JSONPath support for $.property
                if (std.mem.startsWith(u8, request_var.path, "$.")) {
                    const property = request_var.path[2..];
                    return extractJsonProperty(allocator, body, property);
                }

                // If no special path, return whole body
                return try allocator.dupe(u8, body);
            }
            return null;
        },
        .headers => {
            if (result.response_headers) |headers| {
                for (headers) |header| {
                    if (std.ascii.eqlIgnoreCase(header.name, request_var.path)) {
                        return try allocator.dupe(u8, header.value);
                    }
                }
            }
            return null;
        },
    }
}

fn extractJsonProperty(allocator: Allocator, json_body: []const u8, property: []const u8) !?[]const u8 {
    // Handle nested properties like "json.token"
    var parts = std.mem.splitSequence(u8, property, ".");
    var current_json = json_body;
    var temp_json: ?[]const u8 = null;
    defer if (temp_json) |t| allocator.free(t);

    while (parts.next()) |part| {
        const result = extractSimpleJsonProperty(allocator, current_json, part) catch |err| {
            return err;
        };
        if (result) |value| {
            if (temp_json) |t| allocator.free(t);
            temp_json = value;
            current_json = value;
        } else {
            return null;
        }
    }

    if (temp_json) |value| {
        return try allocator.dupe(u8, value);
    }

    return null;
}

fn extractSimpleJsonProperty(allocator: Allocator, json_body: []const u8, property: []const u8) !?[]const u8 {
    // Simple JSON property extraction
    // Look for "property": "value" or "property": value

    var search_pattern = std.ArrayList(u8).init(allocator);
    defer search_pattern.deinit();

    try search_pattern.appendSlice("\"");
    try search_pattern.appendSlice(property);
    try search_pattern.appendSlice("\":");

    if (std.mem.indexOf(u8, json_body, search_pattern.items)) |start_pos| {
        var pos = start_pos + search_pattern.items.len;

        // Skip whitespace
        while (pos < json_body.len and (json_body[pos] == ' ' or json_body[pos] == '\t' or json_body[pos] == '\n' or json_body[pos] == '\r')) {
            pos += 1;
        }

        if (pos >= json_body.len) return null;

        // Check if value is a string (starts with ")
        if (json_body[pos] == '"') {
            pos += 1; // Skip opening quote
            const value_start = pos;

            // Find closing quote (basic implementation, doesn't handle escaped quotes)
            while (pos < json_body.len and json_body[pos] != '"') {
                pos += 1;
            }

            if (pos < json_body.len) {
                return try allocator.dupe(u8, json_body[value_start..pos]);
            }
        } else if (json_body[pos] == '{') {
            // Handle object value
            var brace_count: i32 = 1;
            pos += 1;
            const value_start = pos - 1;

            while (pos < json_body.len and brace_count > 0) {
                if (json_body[pos] == '{') {
                    brace_count += 1;
                } else if (json_body[pos] == '}') {
                    brace_count -= 1;
                }
                pos += 1;
            }

            if (brace_count == 0) {
                return try allocator.dupe(u8, json_body[value_start..pos]);
            }
        } else {
            // Non-string value (number, boolean, null)
            const value_start = pos;

            // Find end of value (comma, }, ], or whitespace)
            while (pos < json_body.len and
                json_body[pos] != ',' and
                json_body[pos] != '}' and
                json_body[pos] != ']' and
                json_body[pos] != ' ' and
                json_body[pos] != '\t' and
                json_body[pos] != '\n' and
                json_body[pos] != '\r')
            {
                pos += 1;
            }

            return try allocator.dupe(u8, json_body[value_start..pos]);
        }
    }

    return null;
}

pub fn findRequestVariables(allocator: Allocator, input: []const u8) !std.ArrayList(RequestVariable) {
    var variables = std.ArrayList(RequestVariable).init(allocator);

    var i: usize = 0;
    while (i < input.len) {
        if (i + 1 < input.len and input[i] == '{' and input[i + 1] == '{') {
            var j = i + 2;
            while (j + 1 < input.len and !(input[j] == '}' and input[j + 1] == '}')) {
                j += 1;
            }

            if (j + 1 < input.len) {
                const var_ref = input[i .. j + 2];

                // Check if this looks like a request variable (contains dots)
                if (std.mem.count(u8, var_ref, ".") >= 3) {
                    const request_var = parseRequestVariable(allocator, var_ref) catch {
                        i = j + 2;
                        continue;
                    };
                    try variables.append(request_var);
                }

                i = j + 2;
            } else {
                i += 1;
            }
        } else {
            i += 1;
        }
    }

    return variables;
}

pub fn substituteRequestVariables(allocator: Allocator, input: []const u8, context: []const RequestContext) ![]const u8 {
    var result = std.ArrayList(u8).init(allocator);
    defer result.deinit();

    var i: usize = 0;
    while (i < input.len) {
        if (i + 1 < input.len and input[i] == '{' and input[i + 1] == '{') {
            var j = i + 2;
            while (j + 1 < input.len and !(input[j] == '}' and input[j + 1] == '}')) {
                j += 1;
            }

            if (j + 1 < input.len) {
                const var_ref = input[i .. j + 2];

                // Check if this looks like a request variable (contains dots)
                if (std.mem.count(u8, var_ref, ".") >= 3) {
                    const request_var = parseRequestVariable(allocator, var_ref) catch {
                        // If parsing fails, keep original text
                        try result.appendSlice(var_ref);
                        i = j + 2;
                        continue;
                    };
                    defer request_var.deinit(allocator);

                    const value = extractRequestVariableValue(allocator, request_var, context) catch {
                        // If extraction fails, keep original text
                        try result.appendSlice(var_ref);
                        i = j + 2;
                        continue;
                    };

                    if (value) |val| {
                        try result.appendSlice(val);
                        allocator.free(val);
                    } else {
                        // If value not found, keep original text
                        try result.appendSlice(var_ref);
                    }
                } else {
                    // Not a request variable, keep original text
                    try result.appendSlice(var_ref);
                }

                i = j + 2;
            } else {
                try result.append(input[i]);
                i += 1;
            }
        } else {
            try result.append(input[i]);
            i += 1;
        }
    }

    return try result.toOwnedSlice();
}

File: src/request_variables.rs
use crate::types::{RequestContext, RequestVariable, RequestVariableSource, RequestVariableTarget};
use anyhow::{anyhow, Result};

pub fn parse_request_variable(reference: &str) -> Result<RequestVariable> {
    // Parse syntax: {{<request_name>.(request|response).(body|headers).(*|JSONPath|XPath|<header_name>)}}

    let mut cleaned = reference;
    if reference.starts_with("{{") && reference.ends_with("}}") {
        cleaned = &reference[2..reference.len() - 2];
    }

    let parts: Vec<&str> = cleaned.split('.').collect();
    if parts.len() < 4 {
        return Err(anyhow!("Invalid request variable format"));
    }

    let request_name = parts[0];
    let source_str = parts[1];
    let target_str = parts[2];
    let path = parts[3..].join(".");

    let source = match source_str {
        "request" => RequestVariableSource::Request,
        "response" => RequestVariableSource::Response,
        _ => return Err(anyhow!("Invalid source: {}", source_str)),
    };

    let target = match target_str {
        "body" => RequestVariableTarget::Body,
        "headers" => RequestVariableTarget::Headers,
        _ => return Err(anyhow!("Invalid target: {}", target_str)),
    };

    Ok(RequestVariable {
        reference: reference.to_string(),
        request_name: request_name.to_string(),
        source,
        target,
        path,
    })
}

pub fn extract_request_variable_value(
    request_var: &RequestVariable,
    context: &[RequestContext],
) -> Result<Option<String>> {
    // Find the request context by name
    let target_context = context
        .iter()
        .find(|ctx| ctx.name == request_var.request_name);

    if target_context.is_none() {
        return Ok(None);
    }

    let ctx = target_context.unwrap();

    match request_var.source {
        RequestVariableSource::Request => extract_from_request(request_var, &ctx.request),
        RequestVariableSource::Response => {
            if let Some(ref result) = ctx.result {
                extract_from_response(request_var, result)
            } else {
                Ok(None)
            }
        }
    }
}

fn extract_from_request(
    request_var: &RequestVariable,
    request: &crate::types::HttpRequest,
) -> Result<Option<String>> {
    match request_var.target {
        RequestVariableTarget::Body => {
            if let Some(ref body) = request.body {
                Ok(Some(body.clone()))
            } else {
                Ok(None)
            }
        }
        RequestVariableTarget::Headers => {
            for header in &request.headers {
                if header.name.eq_ignore_ascii_case(&request_var.path) {
                    return Ok(Some(header.value.clone()));
                }
            }
            Ok(None)
        }
    }
}

fn extract_from_response(
    request_var: &RequestVariable,
    result: &crate::types::HttpResult,
) -> Result<Option<String>> {
    match request_var.target {
        RequestVariableTarget::Body => {
            if let Some(ref body) = result.response_body {
                if request_var.path == "*" {
                    return Ok(Some(body.clone()));
                }

                // Basic JSONPath support for $.property
                if request_var.path.starts_with("$.") {
                    let property = &request_var.path[2..];
                    return extract_json_property(body, property);
                }

                Ok(Some(body.clone()))
            } else {
                Ok(None)
            }
        }
        RequestVariableTarget::Headers => {
            if let Some(ref headers) = result.response_headers {
                for (name, value) in headers {
                    if name.eq_ignore_ascii_case(&request_var.path) {
                        return Ok(Some(value.clone()));
                    }
                }
            }
            Ok(None)
        }
    }
}

fn extract_json_property(json_body: &str, property: &str) -> Result<Option<String>> {
    // Handle nested properties like "json.token" and array indexing like "data[0].version"
    let parts: Vec<&str> = property.split('.').collect();
    let mut current_json = json_body.to_string();

    for part in parts {
        // Check if this part contains array indexing like "data[0]"
        if let Some(bracket_pos) = part.find('[') {
            let property_name = &part[..bracket_pos];
            let index_part = &part[bracket_pos..];

            // Extract the property first
            match extract_simple_json_property(&current_json, property_name) {
                Ok(Some(value)) => {
                    current_json = value;

                    // Now handle array indexing
                    if let Some(index_value) = parse_array_index(index_part) {
                        match extract_array_element(&current_json, index_value) {
                            Ok(Some(element)) => current_json = element,
                            Ok(None) => return Ok(None),
                            Err(e) => return Err(e),
                        }
                    } else {
                        return Err(anyhow!("Invalid array index: {}", index_part));
                    }
                }
                Ok(None) => return Ok(None),
                Err(e) => return Err(e),
            }
        } else {
            match extract_simple_json_property(&current_json, part) {
                Ok(Some(value)) => current_json = value,
                Ok(None) => return Ok(None),
                Err(e) => return Err(e),
            }
        }
    }

    Ok(Some(current_json))
}

fn parse_array_index(index_str: &str) -> Option<usize> {
    // Parse "[0]" or "[123]" format
    if index_str.starts_with('[') && index_str.ends_with(']') {
        let index_num = &index_str[1..index_str.len() - 1];
        index_num.parse::<usize>().ok()
    } else {
        None
    }
}

fn extract_array_element(json_body: &str, index: usize) -> Result<Option<String>> {
    let trimmed = json_body.trim();

    if !trimmed.starts_with('[') {
        return Err(anyhow!("Expected array but got: {}", trimmed));
    }

    let chars: Vec<char> = trimmed.chars().collect();
    let mut pos = 1; // Skip opening [
    let mut current_index = 0;

    while pos < chars.len() {
        // Skip whitespace
        while pos < chars.len() && matches!(chars[pos], ' ' | '\t' | '\n' | '\r') {
            pos += 1;
        }

        if pos >= chars.len() || chars[pos] == ']' {
            return Ok(None); // Array ended, index not found
        }

        let element_start = pos;

        // Find the end of this array element
        let mut depth = 0;
        let mut in_string = false;
        let mut escape_next = false;

        while pos < chars.len() {
            if escape_next {
                escape_next = false;
                pos += 1;
                continue;
            }

            match chars[pos] {
                '\\' => escape_next = true,
                '"' if !escape_next => in_string = !in_string,
                '{' | '[' if !in_string => depth += 1,
                '}' | ']' if !in_string => depth -= 1,
                ',' if !in_string && depth == 0 => break,
                _ => {}
            }

            if depth < 0 {
                break; // Closing array bracket
            }

            pos += 1;
        }

        if current_index == index {
            let element: String = chars[element_start..pos].iter().collect();
            return Ok(Some(element.trim().to_string()));
        }

        current_index += 1;

        // Skip comma
        if pos < chars.len() && chars[pos] == ',' {
            pos += 1;
        }
    }

    Ok(None)
}

fn extract_simple_json_property(json_body: &str, property: &str) -> Result<Option<String>> {
    let search_pattern = format!("\"{}\":", property);

    if let Some(start_pos) = json_body.find(&search_pattern) {
        let mut pos = start_pos + search_pattern.len();
        let chars: Vec<char> = json_body.chars().collect();

        // Skip whitespace
        while pos < chars.len() && matches!(chars[pos], ' ' | '\t' | '\n' | '\r') {
            pos += 1;
        }

        if pos >= chars.len() {
            return Ok(None);
        }

        // Check if value is a string (starts with ")
        if chars[pos] == '"' {
            pos += 1;
            let value_start = pos;

            // Find closing quote, handling escaped quotes
            let mut escape_next = false;
            while pos < chars.len() {
                if escape_next {
                    escape_next = false;
                } else if chars[pos] == '\\' {
                    escape_next = true;
                } else if chars[pos] == '"' {
                    break;
                }
                pos += 1;
            }

            if pos < chars.len() {
                let value: String = chars[value_start..pos].iter().collect();
                return Ok(Some(value));
            }
        } else if chars[pos] == '{' {
            // Handle object value
            let mut brace_count = 1;
            pos += 1;
            let value_start = pos - 1;

            while pos < chars.len() && brace_count > 0 {
                if chars[pos] == '{' {
                    brace_count += 1;
                } else if chars[pos] == '}' {
                    brace_count -= 1;
                }
                pos += 1;
            }

            if brace_count == 0 {
                let value: String = chars[value_start..pos].iter().collect();
                return Ok(Some(value));
            }
        } else if chars[pos] == '[' {
            // Handle array value
            let mut bracket_count = 1;
            pos += 1;
            let value_start = pos - 1;
            let mut in_string = false;
            let mut escape_next = false;

            while pos < chars.len() && bracket_count > 0 {
                if escape_next {
                    escape_next = false;
                } else if chars[pos] == '\\' {
                    escape_next = true;
                } else if chars[pos] == '"' {
                    in_string = !in_string;
                } else if !in_string {
                    if chars[pos] == '[' {
                        bracket_count += 1;
                    } else if chars[pos] == ']' {
                        bracket_count -= 1;
                    }
                }
                pos += 1;
            }

            if bracket_count == 0 {
                let value: String = chars[value_start..pos].iter().collect();
                return Ok(Some(value));
            }
        } else {
            // Non-string value (number, boolean, null)
            let value_start = pos;

            while pos < chars.len()
                && !matches!(chars[pos], ',' | '}' | ']' | ' ' | '\t' | '\n' | '\r')
            {
                pos += 1;
            }

            let value: String = chars[value_start..pos].iter().collect();
            return Ok(Some(value));
        }
    }

    Ok(None)
}

pub fn substitute_request_variables(input: &str, context: &[RequestContext]) -> Result<String> {
    let mut result = String::new();
    let chars: Vec<char> = input.chars().collect();
    let mut i = 0;

    while i < chars.len() {
        if i + 1 < chars.len() && chars[i] == '{' && chars[i + 1] == '{' {
            let mut j = i + 2;
            while j + 1 < chars.len() && !(chars[j] == '}' && chars[j + 1] == '}') {
                j += 1;
            }

            if j + 1 < chars.len() {
                let var_ref: String = chars[i..=j + 1].iter().collect();

                // Check if this looks like a request variable (contains dots)
                if var_ref.matches('.').count() >= 3 {
                    match parse_request_variable(&var_ref) {
                        Ok(request_var) => {
                            match extract_request_variable_value(&request_var, context) {
                                Ok(Some(value)) => result.push_str(&value),
                                _ => result.push_str(&var_ref),
                            }
                        }
                        Err(_) => result.push_str(&var_ref),
                    }
                } else {
                    result.push_str(&var_ref);
                }

                i = j + 2;
            } else {
                result.push(chars[i]);
                i += 1;
            }
        } else {
            result.push(chars[i]);
            i += 1;
        }
    }

    Ok(result)
}
